xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py

```diff
@@ -0,0 +1,343 @@
+                'is_delete_marker': False,
+                "storage_class": obj.get("StorageClass").title(),
+                "e_tag": obj.get('ETag'),
+                "owner_name": obj['Owner'].get('DisplayName') if obj['Owner'].get('DisplayName') else obj['Owner'].get('ID'),
+                "object_url": "https://{0}.s3.{1}.amazonaws.com/{2}".format(bucket_name, bucket_location ,key.replace(" ", "+")),
+                "actual_size": obj.get("Size"),
+            }
+
+            if key not in versions_by_key:
+                versions_by_key[key] = []
+
+            versions_by_key[key].append(version_dict)
+
+        if delete_markers:
+            for marker in delete_markers:
+                key = marker.get("Key")
+                if key == main_folder:
+                    continue
+                
+                marker_dict = {
+                    'version_id': marker.get('VersionId'),
+                    'is_latest': marker.get('IsLatest', False),
+                    'last_modified': marker.get('LastModified'),
+                    'size': 'Deleted',
+                    'storage_class': None,
+                    'type': 'Delete Marker',
+                    'is_delete_marker': True
+                }
+
+                if key not in versions_by_key:
+                    versions_by_key[key] = []
+
+                versions_by_key[key].append(marker_dict)
+
+        files_list = []
+        for key, versions in versions_by_key.items():
+            item_type = key.split(".")[-1] if "." in key else "Unknown"
+            file_name = key.split("/")[-1] if not flat else key
+            latest_version = next((v for v in versions if v['is_latest']), versions[0])  # get the latest version if available, otherwise use the first version
+            item_type = get_file_type(file_name)  # Use the type of the latest version, or extract from the file name if not available
+            file_dict = {
+                "key": key,
+                "url": key,
+                "name": file_name,
+                "path": key.replace(file_name, ""),
+                "storage_class": versions[0].get('StorageClass'),
+                "item_type": item_type,
+                "s3_uri": "s3://{0}/{1}".format(bucket_name, key),
+                "arn": "arn:aws:s3:::{0}/{1}".format(bucket_name, key),
+                'versions': versions,
+                "item_type": item_type,  # Use the type of the latest version
+                "download_url": latest_version.get('download_url'),  # Use the download_url of the latest version
+                "latest_version": latest_version,
+            }
+
+            files_list.append(file_dict)
+
+
+        return sorted(files_list, key=lambda k: str(k["key"]).lower(), reverse=not sort_a_z)
+
+    except Exception as e:
+        print(
+            "Error on line {}: {} {}".format(
+                sys.exc_info()[-1].tb_lineno, type(e).__name__, e
+            )
+        )
```

Comment:
All of these kinds of print & log statements would be better as logger.exception()

https://docs.python.org/3/library/logging.html#logging.Logger.exception